### PR TITLE
Created an abstraction for the Win32 named pipe.

### PR DESCRIPTION
**Pull request by CarlosNihelton**
Named pipe will be used for console redirection. This pipe technique has been tested with C++ and Dart console apps and Flutter GUI apps to make sure it works. 

A console service will use this class to activate and deactivate redirection according to the state of the launcher application. Handling it thru Win32 API all by itself can become really tricky. This class should facilitate that task.

Tests added assert the logic and invariants of this abstraction, not the pipe functionality.

This component is not yet connected to the launcher application. No change in behavior of the app should be noticed by this addition. Integration will happen after adding all necessary components.

The foreseen usage for this component is like the snippets below:

```cpp
auto console = ConsoleService<LocalNamedPipe>();
```

```cpp
template<class Pipe>
class ConsoleService
{
  private:
    Pipe redirectTo;                                     // Here it is
    ConsoleState previousConsoleState;
    ConsoleState consoleState();

  public:
    const HANDLE redirectConsole();
    bool restoreConsole();
    bool hideConsoleWindow();
    bool showConsoleWindow();
};

const HANDLE ConsoleService::redirectConsole()
  {
      // Saving the standard streams context before redirecting.
      previousConsoleState = consoleState();

      // Overriding the std streams with the handle to the write end of the pipe.
      SetStdHandle(STD_OUTPUT_HANDLE, redirectTo.writeHandle());                  // Here it is
      SetStdHandle(STD_ERROR_HANDLE, redirectTo.writeHandle());                   // Here it is
     
      int fd = redirectTo.writeFileDescriptor();                             // Here it is
      auto res = _dup2(fd, _fileno(stdout));
      auto res2 = _dup2(fd, _fileno(stderr));
      setvbuf(stdout, nullptr, _IONBF, 0);
      setvbuf(stderr, nullptr, _IONBF, 0);
      std::ios::sync_with_stdio();
      redirectTo.closeWriteHandles();                             // Here it is
      return redirectTo.readHandle();                             // Here it is
}
```

### CODE REVIEW

**Comment by CarlosNihelton** [comment]
> ⚠️ Please check out carefully the isses found by `clang-tidy` in the new code ⚠️ ⚠️ `clang-format` found formatting issues in the code submitted.⚠️ Make sure to run clang-format and update this pull request. (1/2)

🤣 Trapped by my own workflow.

**Comment by CarlosNihelton** [comment]
> > ⚠️ Please check out carefully the isses found by `clang-tidy` in the new code ⚠️ ⚠️ `clang-format` found formatting issues in the code submitted.⚠️ Make sure to run clang-format and update this pull request. (1/2)
> 
> 🤣 Trapped by my own workflow.

Now we are good to go.

**Review by didrocks — APPROVED** [review]
Just a few typo. The comments helped me a lot following along, thanks for that!
Excellent work! Feel free to merge once the typos are fixed :)

File: DistroLauncher-Tests/DistroLauncher-Tests/LocalNamedPipeTests2.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#include "stdafx.h"
+#include "gtest/gtest.h"
+// Tests added in a separate file due need to fake win32 APIs.
+
+// This overrides the Win32 CreateFile macro by placing a function with the same signature inside the same namespace as
+// the class will be, thus being favorited by ADT.
+// That's why we don't specify the global namespace when calling the Win32 functions int he application code.
```
[the remark this answers is not in the document]
Comment (CarlosNihelton) [reply DistroLauncher-Tests/DistroLauncher-Tests/LocalNamedPipeTests2.cpp:7]: Now I'm thinking whether I need a new keyboard or a pair of glasses :smile:

[the remark this answers is not in the document]
Comment (didrocks) [reply DistroLauncher-Tests/DistroLauncher-Tests/LocalNamedPipeTests2.cpp]: Always blame the keyboard, ofc! :)

**Pull request merged**